core/src/inputs/advice/merkle_path_set.rs

```
@@ -0,0 +1,318 @@
+    /// Merkle tree of depth 1 has two leaves, depth 2 has four leaves etc.
+    pub fn depth(&self) -> u32 {
+        self.total_depth
+    }
+
+    /// Returns a node at the specified index.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified index not valid for the depth of structure.
+    pub fn get_node(&self, depth: u32, index: u64) -> Result<Word, AdviceSetError> {
+        if index >= 2u64.pow(self.total_depth) {
+            return Err(AdviceSetError::InvalidIndex(self.total_depth, index));
+        }
+        if depth != self.total_depth {
+            return Err(AdviceSetError::InvalidDepth(self.total_depth, depth));
+        }
+
+        let pos = 2u64.pow(depth) + index;
+        let index = (pos / 2) as u64;
+        if is_even(pos) {
+            Ok(self.paths[&index][0])
+        } else {
+            Ok(self.paths[&index][1])
+        }
+    }
+
+    /// Returns a Merkle path to the node at the specified index. The node itself is
+    /// not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified index not valid for the depth of structure.
+    pub fn get_path(&self, depth: u32, index: u64) -> Result<Vec<Word>, AdviceSetError> {
+        if index >= 2u64.pow(self.total_depth) {
+            return Err(AdviceSetError::InvalidIndex(self.total_depth, index));
+        }
+        if depth != self.total_depth {
+            return Err(AdviceSetError::InvalidDepth(self.total_depth, depth));
+        }
+
+        let pos = 2u64.pow(depth) + index;
+        let index = pos / 2;
+
+        let mut path = self.paths[&index].clone();
+
+        if is_even(pos) {
+            path.remove(0);
+            Ok(path)
+        } else {
+            path.remove(1);
+            Ok(path)
+        }
+    }
+
+    /// Replaces the leaf at the specified index with the provided value.
+    ///
+    /// # Errors
+    /// Returns an error if the node with specified index is not in the map.
```

Similar to the above comments: we should probably return an error if the path does not exist in this set.